libraries/eosiolib/core/eosio/crypto_bls_ext.hpp

```diff
@@ -0,0 +1,161 @@
+#pragma once
+
+#include "fixed_bytes.hpp"
+#include "varint.hpp"
+#include "serialize.hpp"
+
+#include <array>
+#include <vector>
+
+namespace eosio {
+
+    namespace internal_use_do_not_use {
+        extern "C" {
+            __attribute__((eosio_wasm_import))
+            int32_t bls_g1_add(const char* op1, uint32_t op1_len, const char* op2, uint32_t op2_len, char* res, uint32_t res_len);
+
+            __attribute__((eosio_wasm_import))
+            int32_t bls_g2_add(const char* op1, uint32_t op1_len, const char* op2, uint32_t op2_len, char* res, uint32_t res_len);
+
+            __attribute__((eosio_wasm_import))
+            int32_t bls_g1_mul(const char* point, uint32_t point_len, const char* scalar, uint32_t scalar_len, char* res, uint32_t res_len);
+
+            __attribute__((eosio_wasm_import))
+            int32_t bls_g2_mul(const char* point, uint32_t point_len, const char* scalar, uint32_t scalar_len, char* res, uint32_t res_len);
+
+            __attribute__((eosio_wasm_import))
+            int32_t bls_g1_exp(const char* points, uint32_t points_len, const char* scalars, uint32_t scalars_len, uint32_t n, char* res, uint32_t res_len);
+
+            __attribute__((eosio_wasm_import))
+            int32_t bls_g2_exp(const char* points, uint32_t points_len, const char* scalars, uint32_t scalars_len, uint32_t n, char* res, uint32_t res_len);
+
+            __attribute__((eosio_wasm_import))
+            int32_t bls_pairing(const char* g1_points, uint32_t g1_points_len, const char* g2_points, uint32_t g2_points_len, uint32_t n, char* res, uint32_t res_len);
+
+            __attribute__((eosio_wasm_import))
+            int32_t bls_g1_map(const char* e, uint32_t e_len, char* res, uint32_t res_len);
+
+            __attribute__((eosio_wasm_import))
+            int32_t bls_g2_map(const char* e, uint32_t e_len, char* res, uint32_t res_len);
+
+            __attribute__((eosio_wasm_import))
+            int32_t bls_fp_mod(const char* s, uint32_t s_len, char* res, uint32_t res_len);
+        }
+    }
+
+    using bls_scalar = std::array<char, 32>;
+    using bls_fp     = std::array<char, 48>;
+    using bls_s      = std::array<char, 64>;
+    using bls_fp2    = std::array<bls_fp, 2>;
+    using bls_g1     = std::array<char, 144>;
+    using bls_g2     = std::array<char, 288>;
```

Comment:
Why represent `bls_fp2` as `std::array<bls_fp, 2>`, but not `bls_g2` as `std::array<bls_f2, 3>` ?

Reply:
@mschoenebeck is better to comment but as per the code `bls_g1` is not used partially same way `bls_fp2` does.
If that will make any sense we will change this as a separate PR.